cpp/ql/test/library-tests/dataflow/taint-tests/stl.h

```diff
@@ -300,4 +306,158 @@ namespace std {
 	template<typename T, class... Args> unique_ptr<T> make_unique(Args&&...);
 
 	template<typename T, class... Args> shared_ptr<T> make_shared(Args&&...);
-}
\ No newline at end of file
+}
+
+// --- pair ---
+
+namespace std {
+	template <class T1, class T2>
+	struct pair {
+		typedef T1 first_type;
+		typedef T2 second_type;
+
+		T1 first;
+		T2 second;
+		pair();
+		pair(const T1& x, const T2& y);
+		template<class U, class V> pair(const pair<U, V> &p);
+
+		void swap(pair& p) /*noexcept(...)*/;
+	};
+
+	template <class T1, class T2> pair<T1,T2> make_pair(T1, T2);
```

Comment:
Updated.

Reply:
It's a bit more complicated. The signature is, to a first approximation,

```cpp
template< class T1, class T2 >
std::pair<V1,V2> make_pair( T1&& t, T2&& u );
```

And then there's a remark that `V1` is shorthand for `std::decay<T1>::type` and similarly for `V2`. That's all very beautiful, the way C++ can be beautiful, but it seems to be a bit much to encode for the purpose of this test. In practice, I think the result is that your new signature will always produce pairs of references instead of pairs of values. That could interfere with modelling. But your original signature took values instead of references, and that probably interferes even more.

As a compromise, I suggest this signature, which I think can support all the examples we need and have the same number of reference indirections:

```cpp
template< class T1, class T2 >
std::pair<T1,T2> make_pair(const T1& t, const T2& u );
```

The `const` part is important since that's what allows rvalues as arguments to `make_pair`.

Reply:
Thanks.  I didn't really figure out the significance of `V1` and `V2` and assumed `V1` was roughly the same as `T1`.

I've pushed a commit with your suggestion, and another commit that I think is closer to the standard using an implementation of `remove_reference_t` (but not modelling any of the other `std::decay` features).